### PR TITLE
Adding support for parsing application metadata

### PR DESCRIPTION
**Pull request by mprashant24**

### CODE REVIEW

**Comment by mprashant24** [comment]
Renamed the PersonOfContactMetadata to PointOfContactMetadata

File: src/main/java/com/contrastsecurity/utils/MetadataDeserializer.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
+package com.contrastsecurity.utils;
```
Comment (gmoore) [inline src/main/java/com/contrastsecurity/utils/MetadataDeserializer.java]: This is a pretty good size piece of code. Can you add a few tests around it?

Comment (mprashant24) [reply src/main/java/com/contrastsecurity/utils/MetadataDeserializer.java]: Added test cases for Metadata deserializer.

**Comment by gmoore** [comment]
@mprashant24 This is working well for getting a single application, but getting all applications is not retrieving the metadata. For example, this block of code prints nothing:

```            
            Applications apps = contrastSDK.getApplications(orgUuid);
            for (Application app : apps.getApplications()) {
                System.out.println(app.getName());

                if(app.getMetadataEntities() != null) {
                    List<MetadataEntity> entities = Arrays.asList(app.getMetadataEntities());
                    for(MetadataEntity entity : entities) {
                        System.out.println(entity);
                    }
                }
            }
```
It's weird to have metadata returned in the application entity call but not the applications entity call

Can you add `&expand=metadata` to UrlBuilder.getApplicationsUrl and update any associated tests? I think that'll do the trick.

File: src/test/java/com/contrastsecurity/ContrastSDKTest.java [hunk 2]
```diff
@@ -1,158 +1,188 @@
-package com.contrastsecurity;
```
[the remark this answers is not in the document]
Comment (mprashant24) [reply src/test/java/com/contrastsecurity/ContrastSDKTest.java]: May be auto formatting.

**Comment by mprashant24** [comment]
@gmoore I dont see metadata in applications endpoint 
image

**Comment by mprashant24** [comment]
@gmoore I dont see metadata in applications endpoint
image

**Comment by gmoore** [comment]
It works, it's just not in the docs for some reason. Regardless, we don't pass any expands in the applications list so no need to start now. I've always found that a bit awkward. 

When you sort out the problem with the entire SDKTest showing up as changed, then I can merge.

**Comment by mprashant24** [comment]
@gmoore I have found the issue it was due to my git version, I had a git version which had core.autocurl true. I updated my git and reverted the changes to HEAD-1 and committed my changes again. Please diff it with HEAD^^ for ContrastSDKTest and you will see only new changes.

**Comment by mprashant24** [comment]
image

File: src/test/java/com/contrastsecurity/ContrastSDKTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (gmoore) [reply src/test/java/com/contrastsecurity/ContrastSDKTest.java]: Looking a lot better now. Thanks.

**Comment by gmoore** [comment]
@mprashant24 Did you mean to add 1dd0616 to this PR?

**Comment by mprashant24** [comment]
@gmoore Yes, those changes are added for customer reporting.

**Comment by gmoore** [comment]
It looks like that commit doesn't apply to application metadata. I'd like to approve and merge the application metadata changes before reviewing new work.

**Comment by mprashant24** [comment]
@gmoore reverted. please close this PR so I can open another for yearly trend filtered by application.

**Comment by gmoore** [comment]
@mprashant24 Will do but the CI build is failing: https://travis-ci.org/Contrast-Security-OSS/contrast-sdk-java/builds/610951059?utm_source=github_status&utm_medium=notification

**Comment by mprashant24** [comment]
@gmoore just saw that, I believe I reverted SDKTest too. Fixed and pushed the new change.

**Comment by gmoore** [comment]
@mprashant24 Looking good. Tested this today and it's working well.

Looks like you missed a few spots for person->point conversion. I'll mark them in a new review. When those are fixed, I'll merge and release.

File: src/main/java/com/contrastsecurity/models/MetadataEntity.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    public NumericMetadata getAsNumericMetadata() throws InvalidConversionException{
+        if(type == MetadataType.NUMERIC)
+            return (NumericMetadata) this;
+        else
+            throw new InvalidConversionException(type.toString(), MetadataType.NUMERIC.toString());
+    }
+
+    public PointOfContactMetadata getAsPersonOfContactMetadata() throws InvalidConversionException{
```
Comment (gmoore) [inline src/main/java/com/contrastsecurity/models/MetadataEntity.java]: Rename to `getAsPointOfContactMetadata`

File: src/main/java/com/contrastsecurity/utils/MetadataDeserializer.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        JsonElement type = jsonObject.get("type");
+        if (type != null) {
+            switch (type.getAsString()) {
+                case "STRING":
+                    return context.deserialize(jsonObject,
+                            FreeformMetadata.class);
+                case "NUMERIC":
+                    return context.deserialize(jsonObject,
+                            NumericMetadata.class);
+                case "PERSON_OF_CONTACT":
```
Comment (gmoore) [inline src/main/java/com/contrastsecurity/utils/MetadataDeserializer.java]: Don't change this one as it is the API key sent in JSON

File: src/main/java/com/contrastsecurity/models/MetadataEntity.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
+package com.contrastsecurity.models;
+
+import com.contrastsecurity.exceptions.InvalidConversionException;
+import com.google.gson.annotations.SerializedName;
+
+public abstract class MetadataEntity {
+
+    public enum MetadataType {
+        STRING, NUMERIC, PERSON_OF_CONTACT;
```
Comment (gmoore) [inline src/main/java/com/contrastsecurity/models/MetadataEntity.java]: POINT_OF_CONTACT

**Comment by gmoore** [comment]
Closing this. I'll just finish it here: https://github.com/Contrast-Security-OSS/contrast-sdk-java/pull/35

**Pull request closed without merging**